traffic_ops/traffic_ops_golang/trafficstats/stats_summary.go

```diff
@@ -0,0 +1,127 @@
+package trafficstats
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/jmoiron/sqlx"
+
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+)
+
+// GetStatsSummary handler for getting stats summaries
+func GetStatsSummary(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{}, []string{})
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	lastSummaryDateStr := inf.Params["lastSummaryDate"]
+	if len(lastSummaryDateStr) != 0 { // Perl only checked for existence of query param
+		getLastSummaryDate(w, r, inf)
+		return
+	}
+
+	getStatsSummary(w, r, inf)
+	return
+}
+
+func getLastSummaryDate(w http.ResponseWriter, r *http.Request, inf *api.APIInfo) {
+	queryParamsToSQLCols := map[string]dbhelpers.WhereColumnInfo{
+		"statName": dbhelpers.WhereColumnInfo{"stat_name", nil},
+	}
+	where, _, _, queryValues, errs := dbhelpers.BuildWhereAndOrderByAndPagination(inf.Params, queryParamsToSQLCols)
+	if len(errs) > 0 {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, util.JoinErrs(errs))
+		return
+	}
+	query := selectQuery() + where + " ORDER BY summary_time DESC"
```

Comment:
I can cause an internal server error in this endpoint by passing the `orderby` query parameter. It isn't listed as supported for this type of request to this endpoint, but typically unexpected/undocumented parameters have no effect and when they cause errors like this it can make it hard for the client to figure out what they did wrong. Plus that's technically a client error, because the poor form of their request caused the failure.

I think maybe `dbhelpers.BuildWhereAndOrderByAndPagination` might be overkill for this when you just wanna add an optional `WHERE` and a static `ORDER BY`. The way I see it, you have three options:

- Don't use a `dbhelpers` function and just build the query manually. That's probably what I would do.
- Write a new `dbhelpers` function for "`WHERE`-only" queries as a wrapper around `dbhelpers.BuildWhereAndOrderByAndPagination`. This'll be helpful if you think this problem can/will/should surface in other endpoints
- Manipulate `inf.Params` prior to calling `dbhelpers.BuildWhereAndOrderByAndPagination` such that it doesn't contain pagination or ordering query parameters. I wouldn't block the PR on this, but for my money it's a bad idea if only because it means that adding supported semantics to that function will break this handler in non-obvious ways that we may not catch right away.

Reply:
Wait how could you cause an internal server error? that isnt returned as part of the WHERE return value so there will not be overlap and I am just ignoring it with `_`

Reply:
uhhh....
NO U